shared/lib_battery_dispatch_pvsmoothing_fom.cpp

```diff
@@ -166,43 +166,192 @@ void dispatch_pvsmoothing_front_of_meter_t::dispatch(size_t year,
+            double energyToStoreClipped = 0;
+            if (_P_cliploss_dc.size() > lifetimeIndex + _forecast_hours) {
+                energyToStoreClipped = std::accumulate(_P_cliploss_dc.begin() + lifetimeIndex, _P_cliploss_dc.begin() + lifetimeIndex + _forecast_hours * _steps_per_hour, 0.0) * _dt_hour;
+            }
+
+
+            /*! Energy need to charge the battery (kWh) */
+            double energyNeededToFillBattery = _Battery->energy_to_fill(m_batteryPower->stateOfChargeMax);
+
+
+            // Always Charge if PV is clipping */
+            if (m_batteryPower->canClipCharge && m_batteryPower->powerSystemClipped > 0)
+            {
+                out_power = -m_batteryPower->powerSystemClipped;
+            }
+
+
+            // PV Smoothing algorithm modified for single timestep
+            // forecast period number of resampled pv power outputs at weather file timestep multiplier
+            ssc_number_t pv_power = 0;
+            size_t num_summed = 0;
+            for (size_t i_sampled = 0; i_sampled < m_batt_dispatch_pvs_timestep_multiplier && (lifetimeIndex + i_sampled) < _P_pv_ac.size(); i_sampled++) {
+                pv_power += _P_pv_ac[lifetimeIndex + i_sampled];
+                num_summed++;
+            }
+            // take mean value
+            pv_power = (num_summed > 0) ? pv_power / num_summed : pv_power;
+            // scale by nameplate per ERPI code
+            pv_power = m_batt_dispatch_pvs_nameplate_ac > 0 ? pv_power / m_batt_dispatch_pvs_nameplate_ac : pv_power;
+
+            // forecast energy
+            ssc_number_t forecast_pv_energy = 0;
+
+            for (size_t i_forecast = 0; i_forecast < m_batt_dispatch_pvs_forecast_shift_periods; i_forecast++) {
+                for (size_t i_sampled = 0; i_sampled < m_batt_dispatch_pvs_timestep_multiplier && (lifetimeIndex + i_sampled + i_forecast * m_batt_dispatch_pvs_timestep_multiplier) < _P_pv_ac.size(); i_sampled++) {
+                    forecast_pv_energy += _P_pv_ac[lifetimeIndex + i_sampled + i_forecast * m_batt_dispatch_pvs_timestep_multiplier];
+                }
+            }
+            // scale by nameplate per ERPI code
+            forecast_pv_energy *= _dt_hour;// energy conversion for weather file (_P_pv_ac) timestep
+            forecast_pv_energy = m_batt_dispatch_pvs_nameplate_ac > 0 ? forecast_pv_energy / m_batt_dispatch_pvs_nameplate_ac : forecast_pv_energy;
+
+            m_batt_dispatch_pvs_PV_ramp_interval = pv_power;
+            m_batt_dispatch_pvs_forecast_pv_energy = forecast_pv_energy;
+
+            ssc_number_t  power_to_energy_conversion_factor = m_batt_dispatch_pvs_timestep_multiplier * _dt_hour;
+            ssc_number_t battery_power_terminal = 0;
+            ssc_number_t forecast_power = 0;
+            ssc_number_t previous_power = m_batt_dispatch_pvs_outpower;
+            ssc_number_t battery_soc = m_batt_dispatch_pvs_battsoc;
```

Comment:
Why is this not `_Battery->SOC()`?

Reply:
@brtietz , with battery_soc = _Battery->SOC(), the results are way off:
<img width="1097" alt="image" src="https://user-images.githubusercontent.com/6498311/154370981-4a0fbcde-d8c5-412a-a863-2dadaa633d4c.png">


with battery_soc = _Battery->SOC() /100.0, closer but much worse agreement than using m_batt_dispatch_pvs_battsoc:
<img width="1855" alt="image" src="https://user-images.githubusercontent.com/6498311/154370889-0542f64f-b9b5-4046-b1c2-c950769385d4.png">

This is why I pursued the implementation in the SAM_744_rte branch of ssc

Reply:
We might want to rename some variables! This was a units issue. The algorithm expects batt_soc to have units of scaled energy (hours, I think) instead of percent. So if battery_energy was 0.2, batt_soc at 50% should be 0.1, not 0.5. I corrected this in cefbcf0a865c377040ed3fbc5b8f12a1f1829316

This changes the number of violations in the tests. I believe the new code is more accurate. Here is an example of the old code in the Phoenix_Validation_alloptions test file:

![old_smoothing](https://user-images.githubusercontent.com/5530592/154571740-6ad62df0-f8c7-4335-9704-f25b60529a46.jpg)

In reality, this is a violation, since the SOC goes to zero and the battery does not provide the desired power. However since the PV smoothing algorithm as pre-run, it didn't catch this. The new code does:

![new_smoothing](https://user-images.githubusercontent.com/5530592/154571918-6ccbf5ca-597d-4fda-a7dc-6adbaa0ef70e.jpg)

In fact, its two violations!

I'll get the tests cleaned up shortly, but it looks like we'll be able to get this in the patch!

Reply:
Tested and I went back to the Python code - now the scaling is looking good. Great work!